modules/rostests/kmtests/rtl/RtlStrSafe.c

```diff
@@ -0,0 +1,185 @@
+
+    KmtFreeGuarded(pBuffer);
+    pBuffer = NULL;
+
+    // Note: RtlUnicodeStringPrintf returns STATUS_BUFFER_OVERFLOW here while RtlUnicodeStringPrintfEx returns STATUS_INVALID_PARAMETER!
+    // Documented on MSDN and verified with the Win10 version of ntstrsafe.h
+    RtlInitEmptyUnicodeString(&UsString, NULL, 0);
+    Status = RtlUnicodeStringPrintf(&UsString, FormatStringStrs, L"AAA", L"BBB", L"CCC");
+    ok_eq_hex(Status, STATUS_BUFFER_OVERFLOW);
+
+
+Cleanup:
+    if (pBuffer)
+        KmtFreeGuarded(pBuffer);
+
+    // None of these functions should have crashed.
+    KmtEndSeh(STATUS_SUCCESS);
+}
+
+TESTAPI
+Test_RtlUnicodeStringPrintfEx()
+{
+    NTSTATUS Status;
+    PWSTR pBuffer = NULL;
+    size_t BufferSize;
+    size_t EqualBytes;
+    UNICODE_STRING RemString;
+    UNICODE_STRING UsString;
+    WCHAR FillResult[10];
+
+    RtlFillMemory(FillResult, sizeof(FillResult), 0xAA);
+
+    KmtStartSeh();
+
+    /* STATUS_SUCCESS test, fill behind flag: low-byte as fill character */
+    BufferSize = 16 * sizeof(WCHAR);
+    pBuffer = KmtAllocateGuarded(BufferSize);
+    if (!pBuffer)
+        goto Cleanup;
+
+    RtlInitEmptyUnicodeString(&UsString, pBuffer, BufferSize);
+    RtlInitEmptyUnicodeString(&RemString, NULL, 0);
+
+    Status = RtlUnicodeStringPrintfEx(&UsString, &RemString, STRSAFE_FILL_BEHIND | 0xAA, FormatStringInts, 1, 2, 3);
+    EqualBytes = RtlCompareMemory(UsString.Buffer, FormatStringIntsResult, sizeof(FormatStringIntsResult) - sizeof(WCHAR));
+    ok_eq_hex(Status, STATUS_SUCCESS);
+    ok_eq_size(EqualBytes, sizeof(FormatStringIntsResult) - sizeof(WCHAR));
+    ok_eq_uint(UsString.Length, sizeof(FormatStringIntsResult) - sizeof(WCHAR));
+    ok_eq_uint(UsString.MaximumLength, BufferSize);
+
+    ok_eq_pointer(RemString.Buffer, &UsString.Buffer[UsString.Length / sizeof(WCHAR)]);
+    ok_eq_uint(RemString.Length, 0);
+    ok_eq_uint(RemString.MaximumLength, UsString.MaximumLength - UsString.Length);
+
+    EqualBytes = RtlCompareMemory(RemString.Buffer, FillResult, sizeof(FillResult));
+    ok_eq_size(EqualBytes, sizeof(FillResult));
+
+    KmtFreeGuarded(pBuffer);
+    pBuffer = NULL;
+
+
+    /* STATUS_BUFFER_OVERFLOW test  */
+    BufferSize = 8 * sizeof(WCHAR);
+    pBuffer = KmtAllocateGuarded(BufferSize);
+    if (!pBuffer)
+        goto Cleanup;
+
+    RtlInitEmptyUnicodeString(&UsString, pBuffer, BufferSize);
+    RtlInitEmptyUnicodeString(&RemString, NULL, 0);
+
+    Status = RtlUnicodeStringPrintfEx(&UsString, &RemString, 0, FormatStringStrs, L"AAA", L"BBB", L"CCC");
+    EqualBytes = RtlCompareMemory(UsString.Buffer, L"AAA BBB ", UsString.Length);
+    ok_eq_hex(Status, STATUS_BUFFER_OVERFLOW);
+    ok_eq_size(EqualBytes, UsString.Length);
+    ok_eq_uint(UsString.Length, UsString.MaximumLength);
+
+    ok_eq_pointer(RemString.Buffer, &UsString.Buffer[UsString.Length / sizeof(WCHAR)]);
+    ok_eq_uint(RemString.Length, 0);
+    ok_eq_uint(RemString.MaximumLength, 0);
+
```

Comment:
It would be interesting to test for a null terminator here (fill the buffer with something non-null before the call though)